File: gfg-weekly-coding-contest-76/Max Coins.java


//User function Template for Java
class Solution{
    // Version 1: Heap 
    // Get the maximum coin we can get by all ranges before current time
    // TC: O(nlogn), SC: O(n)
    /*
    public static int maxCoins(int n,int ranges[][])
    {
        Arrays.sort(ranges, ((a, b) -> Arrays.compare(a, b)));
        PriorityQueue<int[]> pq = new PriorityQueue<int[]> ((a, b) -> Integer.compare(a[0], b[0]));
        int maxv = 0, ans = 0;
        for (int[] interval: ranges) {
            while (!pq.isEmpty() && pq.peek()[0] <= interval[0]) {
                maxv = Math.max(maxv, pq.peek()[1]);
                pq.poll();
            }
            ans = Math.max(ans, interval[2] + maxv);
            pq.offer(new int[] {interval[1], interval[2]});
        }
        return ans;
    }
    */
    
    // Version 2: DP
    // Use binary search to find the location before the range
    // We only can search within the ranges before the current range
    // otherwise it will cause wrong answer
    // TC: O(nlogn), SC: O(n)
    public static int maxCoins(int n,int ranges[][])
    {
        Arrays.sort(ranges, 
            ((a, b) -> (a[1] == b[1]? Integer.compare(a[0], b[0]): 
                        Integer.compare(a[1], b[1]))));
        int[] dp = new int[n + 1];
        Arrays.fill(dp, 0);
        int maxv = 0, ans = 0;
        for (int i = 0; i < n; i++) {
            dp[i + 1] = dp[i];
            int l = 0, r = i, m;
            while (l < r) {
                m = l + (r - l) / 2;
                if (ranges[m][1] > ranges[i][0]) {
                    r = m;
                } else {
                    l = m + 1;
                }
            }
            int index = l;
            ans = Math.max(ans, dp[index] + ranges[i][2]);
            dp[i + 1] = Math.max(dp[i + 1], ranges[i][2]);
        }
        return ans;
    }
    
    
}

File: gfg-weekly-coding-contest-76/Max Coins.py
from typing import List


class Solution:
    # Similar question as the below question in Leetcode
    # https://leetcode.com/problems/two-best-non-overlapping-events/
    
    # Version 1: Heap 
    # Get the maximum coin we can get by all ranges before current time
    # TC: O(nlogn), SC: O(n)
    '''
    def maxCoins(self, n : int, ranges : List[List[int]]) -> int:
        # code here
        import heapq
        ranges.sort()
        option = []
        maxv = 0
        ans = 0
        for s, e, v in ranges:
            while option and option[0][0] <= s:
                item = heapq.heappop(option)
                maxv = max(maxv, item[1])
            ans = max(ans, v + maxv)
            heapq.heappush(option, [e, v])
        return ans
    '''
    
    # Version 2: DP
    # Use binary search to find the location before the range
    # We only can search within the ranges before the current range
    # otherwise it will cause wrong answer
    # TC: O(nlogn), SC: O(n)
    def maxCoins(self, n : int, ranges : List[List[int]]) -> int:
        # code here
        from bisect import bisect_left
        ranges.sort(key = lambda x:(x[1], x[0]))
        dp = [0]*(n + 1)
        pre = []
        ans = 0
        for i in range(n):
            #print(i, ranges[i], dp)
            index = bisect_left(pre, ranges[i][0] + 1)
            ans = max(ans, dp[index] + ranges[i][2])
            dp[i + 1] = max(dp[i], ranges[i][2])
            pre.append(ranges[i][1])
        return ans
    
